Support SauceLabs and BrowserStack cloud services

OndraM opened the pull request:
Few improvements in Steward to simplify support of Selenium cloud services (eg. automatic build number detection, port detection etc.). 

Also automatic publishing of results to SauceLabs API (to make the test results available on SauceLabs Dashboard) was added.

As part of the changes option to pass any capability to the Selenium Server was added  (using `--capability=foo:bar` option of `run` command)

OndraM committed:
Use CI Detector to detect current CI environment

OndraM committed:
Fix URL to Selenium status endpoint to make it compatible with SauceLabs

OndraM committed:
BC: pass instance of PHPUnit test to AbstractPublisher::publishResult() to allow eg. access to WebDriver etc. from the publisher

OndraM committed:
Add capability option to run command which allows to pass any capability to WebDriver server

OndraM committed:
When run in CI environment automatically set test metadata to identify build on services like SauceLabs or BrowserStack

OndraM committed:
Use port 80 as default if cloud service like SauceLabs or BrowserStack is given as server URL

OndraM committed:
When SauceLabs server is used, automatically publish test results to SauceLabs API

OndraM committed:
Update changelog

OndraM committed:
Detect cloud services using their real response not only hostname, to detect also connections using Sauce Connect Proxy etc.

jirinovak commented on src/Selenium/SeleniumServerAdapter.php:
```diff
@@ -21,17 +40,48 @@ public function getLastError()
+        $parts = $this->serverUrlParts;
+
+        $serverUrl = '';
+        $serverUrl .= isset($parts['scheme']) ? $parts['scheme'] . '://' : '';
```
`scheme` and `host` are already validated in `parseServerUrl`. This these checks are unnecessary as they cannot be empty.

OndraM replied:
True, as well as port. Thanks.

jirinovak commented on src/Selenium/SeleniumServerAdapter.php:
```diff
@@ -63,6 +112,84 @@ public function isSeleniumServer($seleniumServerUrl)
+     * @param string $host
+     * @return bool
+     */
+    protected function detectCloudServiceByHost($host)
```
`detectKnownCloudServiceByHost` would be imho better name. Current one is little bit confusing for me - it detects any cloud service?

OndraM replied:
I see, had similar thoughts when naming the method and even originally tried name like you propose, however, it was really long then, so I put it only in the description :)

OndraM committed:
If cloud service was detected print it during startup

OndraM committed:
Set test name capability always, not only in CI environment
It is chaper and easier to send it always (even though it is not standardized capability) than detect whether the service supports the capability or not

jirinovak commented on src/Listener/WebDriverListener.php:
```diff
@@ -74,9 +76,30 @@ public function startTest(\PHPUnit_Framework_Test $test)
             ]
         );
 
+        $extraCapabilitiesJson = $config->capability;
```
Unnecessary variable? Or maybe use it in `json_decode` on line 81.

jirinovak commented on src-tests/Process/ProcessSetCreatorTest.php:
```diff
@@ -276,11 +280,39 @@ public function testShouldPropagateCustomOptionsIntoProcess()
+            $this->publisherMock
+        );
+
+        $files = $this->findDummyTests('DummyTest.php'); // find only one file (we don't need more for the test)
```
Maybe `$file`? Or add extra method for searching only one file to avoid extra comment next to each usage.

OndraM replied:
Well, the comment looks a bit unnecessary, so I just removed it :-).
